pkg/repository/network.go

```diff
@@ -70,61 +59,215 @@ func (r *networkRepository) Delete(ctx context.Context, n *Validated[*metal.Netw
 		return nil, err
 	}
 
-	// FIXME delete in ipam with the help of Tx
-
-	err = r.r.ds.Network().Delete(ctx, nw)
+	info, err := r.r.async.NewNetworkDeleteTask(nw.ID)
 	if err != nil {
 		return nil, err
 	}
+	r.r.log.Info("network delete queued", "info", info)
 
 	return nw, nil
 }
 
-func (r *networkRepository) Create(ctx context.Context, rq *Validated[*apiv2.NetworkServiceCreateRequest]) (*metal.Network, error) {
+// NetworkDeleteHandleFn is called async to ensure all dependent entities are deleted
+// Async deletion must be scheduled by async.NewNetworkDeleteTask
+func (r *Store) NetworkDeleteHandleFn(ctx context.Context, t *asynq.Task) error {
+
+	var payload asyncclient.NetworkDeletePayload
+	if err := json.Unmarshal(t.Payload(), &payload); err != nil {
+		return fmt.Errorf("json.Unmarshal failed: %w %w", err, asynq.SkipRetry)
+	}
+	r.log.Info("delete network handler", "uuid", payload.UUID)
+
+	nw, err := r.ds.Network().Get(ctx, payload.UUID)
+	if err != nil && !errorutil.IsNotFound(err) {
+		return err
+	}
+
+	for _, prefix := range nw.Prefixes {
+		_, err = r.ipam.DeletePrefix(ctx, connect.NewRequest(&ipamv1.DeletePrefixRequest{Cidr: prefix.String(), Namespace: nw.Namespace}))
+		if err != nil && !errorutil.IsNotFound(err) {
+			r.log.Error("network release", "error", err)
+			return err
+		}
+	}
+	if nw.Vrf > 0 {
+		err = r.ds.VrfPool().ReleaseUniqueInteger(ctx, nw.Vrf)
+		if err != nil {
+			return fmt.Errorf("unable to release vrf:%d %w", nw.Vrf, err)
+		}
+	}
+
+	err = r.ds.Network().Delete(ctx, nw)
+	if err != nil && !errorutil.IsNotFound(err) {
+		r.log.Error("network delete", "error", err)
+		return err
+	}
 
+	return nil
+}
+
+func (r *networkRepository) Create(ctx context.Context, rq *Validated[*adminv2.NetworkServiceCreateRequest]) (*metal.Network, error) {
 	req := rq.message
 	var (
-		id       string
-		afs      metal.AddressFamilies
-		prefixes metal.Prefixes
+		id          = pointer.SafeDeref(req.Id)
+		name        = pointer.SafeDeref(req.Name)
+		description = pointer.SafeDeref(req.Description)
+		projectId   = pointer.SafeDeref(req.Project)
+		partition   = pointer.SafeDeref(req.Partition)
+		labels      map[string]string
+		vrf         uint
+
+		nat          bool
+		natType      *metal.NATType
+		privateSuper bool
+		shared       bool
+		underlay     bool
 	)
-	for _, p := range req.Prefixes {
-		parsed, err := netip.ParsePrefix(p)
+
+	if req.Labels != nil {
+		labels = req.Labels.Labels
+	}
+
+	prefixes, err := metal.NewPrefixesFromCIDRs(req.Prefixes)
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	switch req.Type {
+	case apiv2.NetworkType_NETWORK_TYPE_CHILD, apiv2.NetworkType_NETWORK_TYPE_CHILD_SHARED:
+		childPrefixes, parent, err := r.allocateChildPrefixes(ctx, req.Project, req.ParentNetworkId, req.Partition, req.Length, req.AddressFamily)
```

Comment:
I think the creation should also be async as here we have the potential to leak resources.

Potentially task aggregation would also be nice (https://github.com/hibiken/asynq/wiki/Task-aggregation), so we can have small subtasks for every non-atomic backend transaction:

1. Ensure child prefixes in IPAM
1. Acquire VRF
1. Add to metal-api datastore